### ISSUE TITLE
Implementing bitflips in training

### ISSUE THREAD

**Issue by ivicii** [pytorchfi/pytorchfi#39]
Hi 
i am working on my thesis(Master student electrical engineering) and my goal is to make a neural network from an algorithm which can be implemented on a chip.
On hardware also bitflips can occur due to lower voltage/ temperure etc. So I want to improve my neural network by training it with bitflips. 

I read your paper about pytorch FI and you are talking about implementing bitflips during training to improve the performance. This must be on the feedforward part. So here is a part of the training code. I use adam optimizer

`for epoch in range(num_epochs):
    running_loss = 0
    for i, (inputs, labels) in enumerate(train_loader):
        optimizer.zero_grad()
        outputs = net(inputs)
        loss = criterion(outputs, labels)

        loss.backward()
        optimizer.step()`

It was not totally clear where and how to implement. 
Is it correct if I do a bitflip on a neuron or weight on the model in between the loss calculation and loss.backward()?

thanks in advance!

**Comment by ma3mool** [comment]
Hi! 

Here is the code we used for the DSML paper regarding training: https://github.com/jose-sv/pytorchfi_train/blob/master/main.py#L151

Note that it was for an older version of PyTorchFI (specifically, v0.1.91), but the general logic is the same as it was before. The main idea is that we introduce error during the forward pass, and allow backpropogation to "self-correct" to improve training accuracy.

Hope this helps!

**Comment by ivicii** [comment]
hmm i have a really small neural network with 1 hidden layer Relu.
Just with numerical data input and numerical data output.

Do you think your tool will work with my neural network?

**Comment by ma3mool** [comment]
Yes, it should. Keep an eye on #41 #42 and #43

additional support for training with error "out-of-the-box" is in our longer term plan.
